contracts/strategies/traderjoe/TraderJoe.sol

```diff
@@ -245,41 +373,99 @@ contract TraderJoe is UUPSUpgradeable, StrategyOwnablePausableBaseUpgradeable {
+        assetValuations = new Valuation[](binsAmount);
+
+        for (uint256 i; i < binsAmount; ++i) {
+            uint256 lpTokenBalance = strategyStorage.lbPair.balanceOf(
+                address(this),
+                strategyStorage.binIds[i]
+            );
+
+            (uint256 reserveX, uint256 reserveY) = strategyStorage
+                .lbPair
+                .getBin(uint24(strategyStorage.binIds[i]));
+
+            uint256 totalSupply = strategyStorage.lbPair.totalSupply(
+                strategyStorage.binIds[i]
+            );
+
+            uint256 valuation = (lpTokenBalance * (reserveX + reserveY)) /
+                totalSupply;
+
+            assetValuations[i] = Valuation(
+                address(strategyStorage.lbPair),
+                valuation
+            );
+        }
     }
 
     function _getLiabilityValuations(bool, bool)
         internal
         view
         virtual
         override
-        returns (Valuation[] memory liabilityValuations)
+        returns (Valuation[] memory)
     {}
 
-    function getTraderJoeLpBalance() public view returns (uint256) {
+    function adjustBins(
+        uint256[] calldata binIds,
+        uint256[] calldata binAllocationsX,
+        uint256[] calldata binAllocationsY,
+        NameValuePair[] calldata params
+    ) public onlyOwner {
+        checkBinIds(binIds);
+        checkBinAllocations(binAllocationsX);
+        checkBinAllocations(binAllocationsY);
+
         TraderJoeStorage storage strategyStorage = TraderJoeStorageLib
             .getStorage();
 
-        return
-            strategyStorage
-                .masterChef
-                .userInfo(strategyStorage.farmId, address(this))
-                .amount;
+        // Withdraw from all bins.
+        uint256 depositTokenBefore = depositToken.balanceOf(address(this));
+
+        _withdraw(getInvestmentTokenSupply(), params);
+
+        uint256 depositTokenAfter = depositToken.balanceOf(address(this));
+
+        // Set bin IDs and allocations to the given ones.
+        strategyStorage.binIds = binIds;
+        strategyStorage.binAllocationsX = binAllocationsX;
+        strategyStorage.binAllocationsY = binAllocationsY;
+
+        // Deposit into the new bins with the new allocations.
+        uint256 depositTokenIncrement = depositTokenAfter - depositTokenBefore;
+
+        _deposit(depositTokenIncrement, params);
     }
 
-    function getTraderJoeLpReserve() public view returns (uint256) {
-        TraderJoeStorage storage strategyStorage = TraderJoeStorageLib
-            .getStorage();
+    function checkBinIds(uint256[] calldata binIds) private pure {
+        uint256 binsAmount = binIds.length;
+
+        require(binsAmount >= 1, "TraderJoe: too few bins");
+        require(binsAmount <= 51, "TraderJoe: too many bins");
 
-        (uint256 reserve0, uint256 reserve1, ) = strategyStorage
-            .lpToken
-            .getReserves();
+        for (uint256 i; i < binsAmount; i++) {
```

Comment:
it is a bit strange looking that we don't initialize i and we rely on the default value, any reason for it? I haven't seen this in any other protocol.

Reply:
Cuz it's unnecessary. TJ V2 uses it fyi

Reply:
ok